lib/redis/connection/ruby.rb

```diff
@@ -34,20 +35,47 @@ def write_timeout=(timeout)
       end
 
       def read(nbytes)
-        result = @buffer.slice!(0, nbytes)
+        result = @buffer.byteslice(@offset, nbytes)
+        @offset += result.bytesize
 
         buffer = String.new(capacity: nbytes, encoding: Encoding::ASCII_8BIT)
         result << _read_from_socket(nbytes - result.bytesize, buffer) while result.bytesize < nbytes
 
         result
       end
 
+      def getbyte
+        fill_buffer if empty_buffer?
+
+        @buffer[@offset].tap do
+          @offset += 1
+        end
+      end
+
       def gets
-        while (crlf = @buffer.index(CRLF)).nil?
-          @buffer << _read_from_socket(16_384)
+        while (crlf = @buffer.index(CRLF, @offset)).nil?
+          fill_buffer
+        end
+
+        end_index = crlf + CRLF.bytesize - 1
+        @buffer.byteslice(@offset..end_index).tap do
+          @offset = end_index + 1
         end
+      end
+
+      def fill_buffer
+        data = _read_from_socket(16_384)
+
+        if empty_buffer?
+          @offset = 0
+          @buffer = data
+        else
+          @buffer << data
+        end
+      end
```

Comment:
Why not ?: 
```suggestion
        if empty_buffer?
          @offset = 0
          _read_from_socket(16_384, @buffer)
        else
          @buffer << _read_from_socket(16_384)
        end
      end
```

Reply:
I think this would fail in JRuby where the buffer is just discarded.

Reply:
I don't know what you mean. Can you elaborate?

Reply:
`_read_from_socket` in JRuby just ignores the 2nd argument: https://github.com/redis/redis-rb/blob/bee6fa273f09b97d6fac13a912f2292fa0f5ebca/lib/redis/connection/ruby.rb#L165-L171

So it won't actually fill up `@buffer` with the data that was read.

Reply:
Ah, right.

Then:

```ruby
@buffer = _read_from_socket(16_384, @buffer)
```

Reply:
Also that comment should probably be investigated, it's likely a decade old, and JRuby may no longer need a different codepath.

Reply:
I've updated this with your suggestion 👍